_published/0013-authentication.md

```diff
@@ -0,0 +1,171 @@
+
+## Summary
+
+Wharf should have authentication. Wharf currently has no authentication.
+This RFC will outline a plan to implement OpenID Connect (OIDC) in Wharf.
+The plan shall be limited to authentication and authorization with
+OIDC. User and group management will be offloaded onto the Identity provider
+this requires using Identity providers which embrace the concept of
+Roll-Based Access Control. The initial implementation will also be limited
+to only sourcing from 1 identity provider at a time, it will however be
+possible to extend this to multiple different sources later.
+
+## Dictionary
+
+- OIDC - [OpenID Connect (OIDC)](https://openid.net/connect)
+
+- IDP - Identity provider - some examples include
+  [Google Workspace](https://workspace.google.com/intl/sv),
+  [Azure AD](https://azure.microsoft.com/en-us/services/active-directory), and
+  [Auth0/Okta](https://www.okta.com/)
+
+- RBAC - [Roll-Based Access Control](https://auth0.com/docs/authorization/rbac/)
+
+- Issuer - In this RFC this refers to the issuer field in a OIDC metadata
+  sheet. The issuer is an authority for validating JWT tokens used in OIDC.
+
+## Motivation
+
+Some claim a good feature of secure infrastructure is access control. Currently,
+our only solution is network isolation. Being able to restrict access to varying
+degrees is also a product goal to satisfy business needs.
+
+Tying the user management of wharf to central repositories such as
+organisation wide IDPs is good for a number of reasons.
+
+- Centrally managed on/off boarding:
+
+  - Leads to less administration
+
+  - Roll based security inheritance allows automatic updates of user access
+    levels
+
+- Piggybacking on existing password and user recovery routines results in a
+  DRY style system architecture
+
+- Near perfectly mirrored compliance with top level organisational
+  security levels
+
+## Explanation
+
+By using the OIDC standard an authentication & authorisation system will be
+built in a minimalist fashion that offloads most of the work involved in
+administration to other pre-existing platforms.
+
+![Connectivity visualised](../assets/auth-map-v2.PNG)
+
+### The Access Token
+
+A OIDC type access token can be used for authentication. The custom claims
+(requires an IDP with support for RBAC) can be used to specify access levels.
+To keep the scope manageable the task of dividing up and refining access
+levels and permission will be left to a later date. For this RFC all kinds
+of read permissions won't require valid tokens, but all write type requests
+will require valid ID-tokens that have been signed by a pre-approved issuer.
+The OIDC protocol defines a token called an
+[Access Token](https://auth0.com/docs/security/tokens/access-tokens).
+The access token is signed by the source issuer. This signature is then used to
+validity and authenticity of the token. See
```

Comment:
```suggestion
validate the authenticity of the token. See
```
or
```suggestion
check the validity and authenticity of the token. See
```
?

Reply:
I guess techinically the authenticity is gained from the `aud` attribute which has to be checked separately when the whole token has been had its signature validated. Going with  `validate the token.`